### DISCUSSION TITLE
How to analyze library code as well (Ruby in particular)

### DISCUSSION THREAD

**Discussion by 0x73746F70626F74686572696E676D65** [github/codeql#13308] [Q&A]
Hello, I'm looking to write queries that also analyze library code, rather than just the base source code. I want to look for file operations being reached in library code, how can I do this? Reading the documentation states that this isn't possible, but I was wondering if it's possible if I somehow bundle the library source into the base source's directory.

**Comment by aeisenberg** [comment]
In general, codeql will analyze everything that gets [_extracted_ into its database](https://docs.github.com/en/code-security/codeql-cli/using-the-codeql-cli/creating-codeql-databases). So, all you need to do is just ensure that the library sources are included in the extraction along with your own local sources. 

This is much easier for interpreted languages than it is for compiled ones. For interpreted languages, you just need to put all of the sources in the same root folder and run extraction from there.

For compiled languages, you will need to use [indirect build tracing](https://docs.github.com/en/code-security/codeql-cli/using-the-codeql-cli/creating-codeql-databases#using-indirect-build-tracing).
